src/distPlus/symbolic/SymbolicDist.re

```diff
@@ -296,103 +313,407 @@ module GenericSimple = {
+      })
+    |> Distributions.Continuous.reduce((+.));
+
+    let sc1d = downsample(sc1);
+    let sc2d = downsample(sc2);
+
+    let ccxy = jsContinuousCombinationConvolve(sc1d.xyShape.xs, sc1d.xyShape.ys, sc2d.xyShape.xs, sc2d.xyShape.ys, func) |> combinePointConvolutionResults;
+    let dcxy = jsContinuousCombinationConvolve(sc1d.xyShape.xs, sc1d.xyShape.ys, sc2d.xyShape.xs, sc2d.xyShape.ys, func) |> combinePointConvolutionResults;
+    let cdxy = jsContinuousCombinationConvolve(sc1d.xyShape.xs, sc1d.xyShape.ys, sc2d.xyShape.xs, sc2d.xyShape.ys, func) |> combinePointConvolutionResults;
+    let continuousShapeSum = Distributions.Continuous.reduce((+.), [|ccxy, dcxy, cdxy|]);
+
+    (continuousShapeSum, ddxy)
   };
 
-  let toShape = (dists: t, sampleCount: int) => {
-    let normalized = normalizeWeights(dists);
-    let continuous =
-      normalized
-      |> E.A.filter(((r, _)) => GenericSimple.contType(r) == `Continuous)
-      |> continuousShape(_, sampleCount);
-    let discrete =
-      normalized
-      |> E.A.filter(((r, _)) => GenericSimple.contType(r) == `Discrete)
-      |> discreteShape(_, sampleCount);
-    let shape =
-      MixedShapeBuilder.buildSimple(~continuous=Some(continuous), ~discrete);
-    shape |> E.O.toExt("");
+  let evaluateCombinationDistribution = (et1: nodeResult, et2: nodeResult, op: operation, n: int) => {
+    /* return either a Distribution or a RenderedShape. Must integrate to 1. */
+
+    let func = funcFromOp(op);
+    switch ((et1, et2, op)) {
+    /* Known cases: replace symbolic with symbolic distribution */
+    | (`Simple(`Float(v1)), `Simple(`Float(v2)), _) => {
+        `Simple(`Float(func(v1, v2)))
+      }
+
+    | (`Simple(`Normal(n2)), `Simple(`Float(v1)), `AddOperation)
+    | (`Simple(`Float(v1)), `Simple(`Normal(n2)), `AddOperation) => {
+        let n: normal = {mean: v1 +. n2.mean, stdev: n2.stdev};
+        `Simple(`Normal(n))
+      }
+
+    | (`Simple(`Normal(n1)), `Simple(`Normal(n2)), `AddOperation) => {
+        let n: normal = {mean: n1.mean +. n2.mean, stdev: sqrt(n1.stdev ** 2. +. n2.stdev ** 2.)};
+        `Simple(`Normal(n));
+      }
+
+    | (`Simple(`Normal(n1)), `Simple(`Normal(n2)), `SubtractOperation) => {
+        let n: normal = {mean: n1.mean -. n2.mean, stdev: sqrt(n1.stdev ** 2. +. n2.stdev ** 2.)};
+        `Simple(`Normal(n));
+      }
+
+    | (`Simple(`Lognormal(l1)), `Simple(`Lognormal(l2)), `MultiplyOperation) => {
+        let l: lognormal = {mu: l1.mu +. l2.mu, sigma: l1.sigma +. l2.sigma};
+        `Simple(`Lognormal(l));
+      }
+
+    | (`Simple(`Lognormal(l1)), `Simple(`Lognormal(l2)), `DivideOperation) => {
+        let l: lognormal = {mu: l1.mu -. l2.mu, sigma: l1.sigma +. l2.sigma};
+        `Simple(`Lognormal(l));
+      }
+
+
+    /* General cases: convolve the XYShapes */
+    | (`Simple(d1), `Simple(d2), _) => {
+        let (sc1, sd1) = renderDistributionToXYShape(d1, n);
+        let (sc2, sd2) = renderDistributionToXYShape(d2, n);
```

Comment:
I'm not sure if you can really render both separately. For instance, if I had ``normal(5,2)`` and ``normal(500, 2)``, and wanted to take the pointwise multiplication, I would want to evaluate them on all of their shared x coordinates and do the calculations right for these coordinates. If I first rendered them, it's possible that each would only have a narrow range of x coordinates that wouldn't overlap, so we'd have to make messy assumptions there.

Reply:
Absolutely, and I don't know of a good way to solve this in general. The most robust way I can think of is to a) use analytic solutions whenever we can, at least for the bounds, and at least approximately, and b) keep the symbolic expressions around, so that we can combine them into a JS string and use that to sample and/or render within those bounds.

But how often *do* users pointwise-multiply distributions like that? I can't think of any situations in which that would be meaningful :thinking:

Reply:
I'm just suggesting that you go through:
1) Figure out the Xs necessary for both distribution. 
2) Find the Ys for each distribution, that correspond to the Xs combined of both distributions; not just 1 distribution.

This is how we currently handle the multimodals, I believe.

Reply:
But does that help? Evaluating `normal(5,2) .* normal(500,2)` on the whole union of the Xs would give a more accurate picture of the tails but still miss the bulk of the distribution.

Alternatively, we could create a new set of evenly spaced Xs based on a heuristically chosen minX and maxX. Could be the extrema of the combined Xs, or something more clever (e.g. `normal(5,2) .* normal(0.5, 0.2)` would really only need to be evaluated on the min/max of the second one).

Reply:
I think it would help for things like ``normal(5,2) .+ normal(500,2)``, which multimodal() kind of is, but you're right that for multiplication it would be tough. 

Evenly spaces Xs is a bit scary to me, because it would do poorly in cases like, ``normal(5,2) .+ normal(500,2)``.

I imagine a more sophisticated system could look a bit like:
- We do a mix of "evenly spaced" and "integral-based-spacing". 
- Once we do one rendering, we see where the probability mass is. Then we do a second rendering where the choices of x values are weighted by the probability mass in this rendering.

Ideally it may even be better to go all out and do MCMC or something. That would come much later though.